pkg/cmd/pulumi/config/io.go

```diff
@@ -321,20 +320,22 @@ func parseKeyValuePair(pair string, path bool) (config.Key, string, error) {
 // (namespace: <project-name>, key: foo.bar:buzz) if path.
 func ParseConfigKey(key string, path bool) (config.Key, error) {
 	// If the key is a path, the namespacing requirement only applies to the
-	// top-level key, while sub-keys may have arbitrary names. Parse the path
-	// and then calculate the namespace for the top-level segment.
+	// top-level key, while sub-keys may have arbitrary names.
 	if path {
-		path, err := resource.ParsePropertyPathStrict(key)
-		if err != nil {
-			// For backwards compatibility, fallback to non-path parsing when key is not a valid path.
-			return ParseConfigKey(key, false)
-		}
-		topLevelKey, err := ParseConfigKey(path[0].(string), false)
-		if err != nil {
-			return config.Key{}, err
+		// Figure out if the key has multiple segments (delimetered by dots and square brackets).
+		// If they exist, we'll parse the first segment to validate its structure and obtain its
+		// namespace. Then we'll use this namespace with the rest of the key.
+		bracketOrDotIndex := strings.IndexAny(key, "[.")
+		if bracketOrDotIndex > 0 {
+			topSegment := key[:bracketOrDotIndex]
+			topKey, err := config.ParseKey(topSegment)
+			if err != nil {
+				return config.Key{}, err
```

Comment:
Shouldn't we still fall back to non-path parsing in this case as we did before for backwards compatibility?

Reply:
I wasn't able to come up with a test case that would get us in this branch but would be fixed with a fallback. I can add it anyway for safety - was that your motivation? Or maybe you see a test case?

Reply:
Yeah, I was worried we are missing something by removing that fallback.  But I gave it a try, and I think what you have here is actually dealing with all the extra cases I imagined fine.  (There are some test failures though that we should take a look at)

Reply:
So I ended up calling `ParseConfigKey(..., false)` here which preserves the old property of falling back to non-path parsing.

I also added the second commit to the PR which allows passing a workspace mock, so that we could extend the test suite with non-workspaced keys. Please take a look.